### DISCUSSION TITLE
Unexpected push token received from bitwarden server

### DISCUSSION THREAD

**Discussion by VectorZhao** [dani-garcia/vaultwarden#3700] [Q&A]
### Subject of the issue
can not using the new feature: Mobile Client push notification

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.0
* Web-vault version: v2023.5.0
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": true,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": true,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "data/attachments",
  "authenticator_disable_time_drift": false,
  "data_folder": "data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "***************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://*******************",
  "domain_origin": "*****://*******************",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Vaultwarden",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 600000,
  "push_enabled": true,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": false,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": null,
  "smtp_password": null,
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": null,
  "templates_folder": "data/templates",
  "tmp_folder": "data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 3012,
  "yubico_client_id": null,
  "yubico_secret_key": null,
  "yubico_server": null
}
```
</details>

### Steps to reproduce
**docker-compose.yml**
```
version: '3.4'

services:

  vaultwarden:
    image: vaultwarden/server:latest
    restart: always
    ports:
      - '8880:80'
      # - '3012:3012'
    volumes:
      - ./vw-data/:/data/
    environment:
      ADMIN_TOKEN: xxxxxxxxxx
      DOMAIN: https://xxxxxxx.xxx.xxx
      SIGNUPS_ALLOWED: "false"
      # WEBSOCKET_ENABLED: "true"
      PUSH_ENABLED: true
      PUSH_INSTALLATION_ID: xxxxxxxx
      PUSH_INSTALLATION_KEY: xxxxx-xxxxxxx-x-b3xxxxx26-xxxxxxxxxxxxx
```
I downloaded Bitwarden from the Google Play Store, and I reinstalled the application and logged back into Vaultwarden.
I obtained the PUSH_INSTALLATION_ID and PUSH_INSTALLATION_KEY by entering my email  from https://bitwarden.com/host/.
### Expected behaviour
Mobile Client push notification

### Actual behaviour
It does not work.

### Troubleshooting data
**error**

```
vaultwarden  | [2023-07-16 07:16:49.265][vaultwarden::api::core::accounts][ERROR] An error occured while proceeding registration of a device: {"ErrorModel":{"Message":"Unexpected push token received from bitwarden server: error decoding response body: missing field `access_token` at line 1 column 26","Object":"error"},"ExceptionMessage":null,"ExceptionStackTrace":null,"InnerExceptionMessage":null,"Message":"Unexpected push token received from bitwarden server: error decoding response body: missing field `access_token` at line 1 column 26","Object":"error","ValidationErrors":{"":["Unexpected push token received from bitwarden server: error decoding response body: missing field `access_token` at line 1 column 26"]},"error":"","error_description":""}
vaultwarden  | [2023-07-16 07:16:49.265][response][INFO] (put_device_token) PUT /api/devices/identifier/<uuid>/token => 400 Bad Request
```

**Comment by BlackDex** [comment]
@VectorZhao  I think you swapped the PUSH ID and KEY if I look at your `environment` variables.
The ID should have dashes `-` and the KEY does not, while in your paste, it does.

Please swap them and see what happens.

**Reply by VectorZhao** [reply]
Worked for swap them, thank you. 
Sorry for wasting your time.

**Reply by BlackDex** [reply]
np, good for other people who encounter the same issue

**Comment by VectorZhao** [comment]
> @VectorZhao I think you swapped the PUSH ID and KEY if I look at your `environment` variables. The ID should have dashes `-` and the KEY does not, while in your paste, it does.
> 
> Please swap them and see what happens.

**Comment by sadden3194** [comment]
Hi all!

I have the same problem.
I added my new envs:

![Bildschirmfoto 2023-07-31 um 12 23 27](https://github.com/dani-garcia/vaultwarden/assets/17011498/65413122-cdea-45e6-be1e-0cab5d83b5a5)

I change a value in bitwarden via pc and than the logs shows:

![Bildschirmfoto 2023-07-31 um 12 22 49](https://github.com/dani-garcia/vaultwarden/assets/17011498/4172806c-cf16-4f55-8dc9-216c19c43554)


To receive the data I used bitwarden.eu, is that correct?

![Bildschirmfoto 2023-07-31 um 12 25 37](https://github.com/dani-garcia/vaultwarden/assets/17011498/a81b3e20-1a1a-4ae4-83d1-2f41b3a54e93)

**Reply by BlackDex** [reply]
Create a US based key.
EU Based keys do not work on `push.bitwarden.com` and there does not seem to be a `push.bitwarden.eu`.
It might be that we need to use the `.com` for push and `.eu` for the identity, but not sure.

So for now, create a US key, as is also mentioned on the wiki page.

**Comment by sadden3194** [comment]
I did it now. but I get the same issue

**Reply by BlackDex** [reply]
You are sure it currently has the correct new values loaded?
And restarted the instance?

I can't say anything else but it seems to work for me, and it only seems to fail if either the key and id are reversed or incorrect.

**Reply by sadden3194** [reply]
now it works! I tried it again with new values :) Thanks!

**Comment by sadden3194** [comment]
But there is another issue.
sync via push notifications works in the own safe.
But If I create a new entry via pc in a folder, which is assigend to an organization, the push sync doesn't work on the iphones.
I have to refresh it on iphone manually to sync.

But if I create the entry from my iphone (in the folder of organization) the push sync to my pc works directly.
is this a bug?

**Reply by kpfleming** [reply]
No. Mobile push does not work for organizations.

The 'PC' (web vault, web browser extensions, etc.) use a completely different mechanism and are not dependent on push notifications to receive updates.

**Reply by sadden3194** [reply]
why it doesn't work for organzations?
Not implemented yet?

**Reply by BlackDex** [reply]
Isn't supported by Bitwarden.

**Reply by sadden3194** [reply]
Need I websocket for this?

**Reply by BlackDex** [reply]
Web-Sockets are needed for all clients except Mobile client. Mobile clients only work via this Push feature.
If you do not have web-sockets working but do configure the Push relay, it should still work for the mobile clients.

**Comment by sadden3194** [comment]
But If I create an entry on my iphone, my pc get this entry directly.
Is this websocket or push?

**Reply by BlackDex** [reply]
That is done via websockets
